app/scripts/modules/core/pipeline/config/pipelineConfigProvider.ts

```diff
@@ -0,0 +1,202 @@
+import { auto, IServiceProvider, module } from 'angular';
+import { cloneDeep, intersection, memoize } from 'lodash';
+import { $log } from 'ngimport';
+
+
+import { Application } from 'core/application/application.model';
+import { IExecution, IStage } from 'core/domain';
+import { ITriggerTypeConfig, IStageTypeConfig } from './validation/pipelineConfig.validator';
+import { SETTINGS } from 'core/config/settings';
+
+export interface ITransformer {
+  transform: (application: Application, execution: IExecution) => void;
+};
+
+export class PipelineConfigProvider implements IServiceProvider {
+  private $injector: auto.IInjectorService;
+
+  private triggerTypes: ITriggerTypeConfig[] = [];
+  private stageTypes: IStageTypeConfig[] = [];
+  private transformers: ITransformer[] = [];
+
+  constructor() {
+    this.getStageConfig = memoize(this.getStageConfig,
```

Comment:
This will fail on line 151 unless you bind `this` or use an arrow function here.